File: python/19/Remove_Nth_Node_From_End_of_List.py
tion for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        if not head or n==0:
            return head
        newhead = ListNode(0)
        newhead.next = head
        p = pre = newhead
        for i in range(n):
            p = p.next
        while p.next:
            pre = pre.next
            p = p.next
        pre.next = pre.next.next
        return newhead.next

File: java/19/RemoveNthNodeFromEndOfList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // 为了避免对删除头指针的情况额外处理，设置哨兵结点
        ListNode preHead = new ListNode(-1);
        preHead.next = head;
        // 快指针比慢指针快n步，这样快指针指向最后一个结点时，慢指针指向要删除的前一个结点
        ListNode slow = preHead;
        ListNode fast = preHead;
        while (n > 0) {
            if (fast.next == null) {
                return preHead.next;
            }
            fast = fast.next;
            n--;
        }

        // 使快指针指向最后一个结点
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
        // 删除结点
        slow.next = slow.next.next;

        return preHead.next;
    }
}
